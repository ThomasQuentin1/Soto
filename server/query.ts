import * as mysql from "mysql";

export const getPool = () => {
  if (process.env && process.env.TEST && process.env.TEST === "TRUE") {
    console.log("db test");
    return {
      host: "51.11.241.109",
      user: "soto",
      password: "s0t0lefeu!",
      database: "algo",
    };
  } else {
    console.log("db prod");
    return {
      host: "51.11.241.109",
      user: "soto",
      password: "s0t0lefeu!",
      database: "users",
    };
  }
};
const algoPool = mysql.createPool({
  host: "51.11.241.109",
  user: "soto",
  password: "s0t0lefeu!",
  database: "algo",
});

const usersPool = mysql.createPool(getPool());

export const algoQuery = async <T>(query: string, values?: string[]) => {
  return new Promise<T[]>((resolve, reject) => {
    algoPool.query({ sql: query, values }, (err: any, results: any) => {
      if (err) reject(err);
      resolve(results);
    });
  });
};

export const usersQuery = async <T>(query: string, values?: string[]) => {
  return new Promise<T[]>((resolve, reject) => {
    usersPool.query({ sql: query, values }, (err: any, results: any) => {
      if (err) reject(err);
      resolve(results);
    });
  });
};

let connections = 0;

export const openConnection = () => {
  connections += 1;
};

export const endConnection = async () =>
  new Promise<void>((resolve) => {
    if (connections != 0) {
      connections -= 1;
      resolve();
    } else {
      setTimeout(() => {
        algoPool.end();
        usersPool.end();
        resolve();
      }, 1000);
    }
  });
