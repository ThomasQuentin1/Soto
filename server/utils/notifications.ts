import { usersQuery } from "../query";

const app_id = "4384c64d-c68b-4dba-ae9d-e05bfc6c21e9";

export const sendNotificationToAllCustomer = async (message: string) => {
  const res = await fetch("https://onesignal.com/api/v1/notifications", {
    method: "POST",
    body: JSON.stringify({
      app_id,
      contents: { en: message },
      url: ``,
    }),
    headers: { "content-type": "application/json" },
  });

  console.log(await res.json());
};

export const sendNotificationToOneCustomer = async (
  message: string,
  customerMail: string
) => {
  const user: any[] = await usersQuery("SELECT * FROM users WHERE email = ?", [
    customerMail,
  ]);
  if (!user[0].pushToken) {
    console.log("no user with this mail");
    return;
  }
  const res = await fetch("https://onesignal.com/api/v1/notifications", {
    method: "POST",
    body: JSON.stringify({
      app_id,
      include_player_ids: [user[0].pushToken],
      contents: { en: message },
      url: ``,
    }),
    headers: { "content-type": "application/json" },
  });

  console.log(await res.json());
};
