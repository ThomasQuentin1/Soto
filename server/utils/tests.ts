import { usersQuery } from "../query";
import resolvers from "../resolvers";
import * as jwt from "jsonwebtoken"

const QorM = async (
  op: "Query" | "Mutation",
  resolver: string,
  args: any,
  token?: string
) => {
  let user;
  try {
    if (token) {
      const {id} = jwt.verify(token, 's0t0') as {id: string};
      const userQ = await usersQuery(
        "SELECT * FROM users WHERE id = ? LIMIT 1",
        [id]
      );
      if (userQ.length == 1) user = userQ[0];
    }
  } catch {}
  
  // @ts-ignore
  const elem: any = resolvers![op]![resolver];
  if (!elem) throw "bad resolver name, or missmatch on query/mutation";
  return await elem({}, args, { user }, {});
};

export const Mutate = async (resolver: string, args: any, token?: string) =>
  QorM("Mutation", resolver, args, token);
export const Query = async (resolver: string, args: any, token?: string) =>
  QorM("Query", resolver, args, token);

export function makeid(length) {
  var result = "";
  var characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  var charactersLength = characters.length;
  for (var i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
}
