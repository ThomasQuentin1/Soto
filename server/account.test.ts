import { AuthenticationError, UserInputError } from "apollo-server-micro";
import { endConnections } from "./query";
import { Mutate, Query } from "./utils/tests";


let token: string = "";
const email = `user${Math.floor(Math.random() * 1000)}@test.com`;

afterAll(() => {
  endConnections();
})


describe('Basic errors cases with account', () => {
  it('should deny access to account without being logged in', async () => {
    await expect(Query("account", {})).rejects.toStrictEqual(new AuthenticationError("please login"));
  })

  it('should deny access with invalid username or password', async () => {
    await expect(Mutate("login", { email: "invalid-email@test.com", passwordSHA256: "invalidPassword123" })).rejects.toStrictEqual(new AuthenticationError("Invalid email or password"));
  })

  it('should deny actions to a unlogged user', async () => {
    await expect(Mutate("setObligations", { obligations: [] })).rejects.toStrictEqual(new AuthenticationError("please login"));
    await expect(Mutate("setCriterions", { criterions: [] })).rejects.toStrictEqual(new AuthenticationError("please login"));
    await expect(Mutate("changeEmail", { newEmail: "newemail" })).rejects.toStrictEqual(new AuthenticationError("please login"));
    await expect(Mutate("changePassword", { newPasswordSHA256: "newpass" })).rejects.toStrictEqual(new AuthenticationError("please login"));
    await expect(Mutate("subscribeNotifications", { token: "tok" })).rejects.toStrictEqual(new AuthenticationError("please login"));
  })

  it('should deny the creation of a double account', async () => {

    const duplicateEmail = `user-dup-${Math.floor(Math.random() * 1000)}@test.com`;
    const token = await Mutate("register", { email: duplicateEmail, passwordSHA256: "rand" });
    await expect(Mutate("register", { email: duplicateEmail, passwordSHA256: "rand" })).rejects.toStrictEqual(new AuthenticationError("Email already in use"));
    await expect(Mutate("removeAccount", { passwordSHA256: "invalid" }, token)).rejects.toStrictEqual(new AuthenticationError("Invalid password"));
    await expect(Mutate("removeAccount", { passwordSHA256: "rand" }, token)).resolves.toBeTruthy();
  })
})

describe('Account', () => {

  beforeAll(async (done) => {
    await Mutate("register", { email, passwordSHA256: "blbl" });
    token = await Mutate("login", { email, passwordSHA256: "blbl" });
    done();
  })

  afterAll(() =>
    Mutate("removeAccount", { passwordSHA256: "blbl" }))

  it('should retrive the user email', async () => {
    const acc = await Query("account", {}, token);
    expect(acc.email).toBe(email);
  })

  it('should save notification token', async () => {
    await Mutate("subscribeNotifications", { token: "thisismytoken" }, token); // should not throw
  })

  it('should change email properly', async () => {
    const oldProfile = await Query("account", {}, token);
    await Mutate("changeEmail", { newEmail: "thisIsMyNewEmail@gmail.com" }, token);
    const newProfile = await Query("account", {}, token);
    expect(oldProfile.email).not.toBe(newProfile.email);
    expect(newProfile.email).toBe("thisIsMyNewEmail@gmail.com");
  })

  it('should change password properly', async () => {
    const tmpemail = `user${Math.floor(Math.random() * 1000)}@test.com`;
    const tmptoken = await Mutate("register", { email: tmpemail, passwordSHA256: "blbl" });
    await Mutate("changePassword", {newPasswordSHA256: "newblbl"}, tmptoken); // This would throw in case of error
    const secondtoken = await Mutate("login", {email : tmpemail, passwordSHA256: "newblbl"});
    expect(secondtoken.length).toBeGreaterThan(1);
    await Mutate("removeAccount", {passwordSHA256: "newblbl"}, secondtoken);
  })

  it('handle obligations properly', async () => {
    const oldProfile = await Query("account", {}, token);
    await Mutate("setObligations", { obligations: [{ id: 1 }] }, token);
    const newProfile = await Query("account", {}, token);
    expect(oldProfile).not.toBe(newProfile);
    expect(newProfile.obligations.find((o: any) => o.activated).id).toBe(1);
    expect(newProfile.obligations.filter((o: any) => o.activated).length).toBe(1);
  })

  it('handle critetions properly', async () => {
    const oldProfile = await Query("account", {}, token);
    await Mutate("setCriterions", { criterions: [{ id: 2, position: 1 }, { id: 1, position: 2 }] }, token);
    const newProfile = await Query("account", {}, token);
    expect(oldProfile).not.toBe(newProfile);
    expect(newProfile.criterions.find((o: any) => o.position == 1).id).toBe(2);
    expect(newProfile.criterions.find((o: any) => o.position == 2).id).toBe(1);
    expect(newProfile.criterions.filter((o: any) => o.activated).length).toBe(2);
  })

  it('sould deny invalid ids into criterions and obligations', async () => {
    await expect(Mutate("setCriterions", { criterions: [{ id: 9999, position: 1 }] }, token)).rejects.toStrictEqual(new UserInputError("Invalid criterion id"));
    await expect(Mutate("setObligations", { obligations: [{ id: 9999 }] }, token)).rejects.toStrictEqual(new UserInputError("Invalid obligation id"));
  })

  it('sould deny invalid positions in criterions', async () => {
    await expect(Mutate("setCriterions", { criterions: [{ id: 1, position: 0 }] }, token)).rejects.toStrictEqual(new UserInputError("Criterions position must start with 1"));
    await expect(Mutate("setCriterions", { criterions: [{ id: 1, position: 1 }, { id: 2, position: 3 }] }, token)).rejects.toStrictEqual(new UserInputError("Criterions position are incorrect (not 1, 2, 3...)"));
  })
})