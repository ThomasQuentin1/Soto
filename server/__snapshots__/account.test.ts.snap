// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop list gathering should retrive 4 leclerc in the area 1`] = `
Array [
  Object {
    "city": "Erstein",
    "code": "056701",
    "id": 1,
    "lat": 48.425563,
    "long": 7.638703,
    "name": "E.Leclerc Drive Erstein",
    "server": "fd12",
  },
  Object {
    "city": "Strasbourg",
    "code": "086702",
    "id": 2,
    "lat": 48.545883,
    "long": 7.767355,
    "name": "E.Leclerc Drive Strasbourg - Neuhof",
    "server": "fd6",
  },
  Object {
    "city": "Strasbourg",
    "code": "086703",
    "id": 3,
    "lat": 48.576761,
    "long": 7.768423,
    "name": "E.Leclerc Drive Strasbourg - Esplanade",
    "server": "fd6",
  },
  Object {
    "city": "Strasbourg",
    "code": "086701",
    "id": 4,
    "lat": 48.597222,
    "long": 7.735704,
    "name": "E.Leclerc Drive Strasbourg March√© Gare",
    "server": "fd6",
  },
]
`;
